keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md

```diff
@@ -195,153 +242,387 @@ import (
-the existing API objects?**
+The external CCMs follow the [same version skew policy][skew] as the kube-controller-manager.
+Providers, and communities, that publish their own CCM are responsible for
+updating those CCMs to follow the version and skew policy.
 
-  No.
+Copied from the [Kubernetes Version Skew Policy documentation][skew]:
 
-* **Will enabling / using this feature result in increasing time taken by any
-operations covered by [existing SLIs/SLOs]?**
+> kube-controller-manager, kube-scheduler, and cloud-controller-manager must not be newer than the kube-apiserver instances they communicate with. They are expected to match the kube-apiserver minor version, but may be up to one minor version older (to allow live upgrades).
 
-  No.
+## Production Readiness Review Questionnaire
 
-* **Will enabling / using this feature result in non-negligible increase of
-resource usage (CPU, RAM, disk, IO, ...) in any components?**
+### Feature Enablement and Rollback
 
-  No. In fact, it should reduce resource usage.
+###### How can this feature be enabled / disabled in a live cluster?
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: DisableCloudProviders
+  - Components depending on the feature gate: kubelet, kube-apiserver, kube-controller-manager
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: DisableKubeletCloudCredentialProvider
+  - Components depending on the feature gate: kubelet
 
-### Troubleshooting
+###### Does enabling the feature change any default behavior?
+Yes, enabling this feature will disable all capabilities enabled when `--cloud-provider` is set in core components.
+Users need to ensure they have migrated to out-of-tree components prior to enabling this feature gate.
+If appropriate extensions (CCM, credential provider, apiserver-network-proxy, etc) are in use, cloud provider capabilities
+should remain the same at the very least.
 
-The Troubleshooting section currently serves the `Playbook` role. We may consider
-splitting it into a dedicated `Playbook` document (potentially with some monitoring
-details). For now, we leave it here.
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+Yes, the feature can be disabled once it is enabled. If disabled, users must ensure
+that the CCM is no longer running in the cluster. Credential provider plugins and the
+apiserver network proxy do not have to be stopped on rollback.
 
-_This section must be completed when targeting beta graduation to a release._
+###### What happens if we reenable the feature if it was previously rolled back?
+All capabilities from in-tree cloud providers will be re-disabled.
 
-* **How does this feature react if the API server and/or etcd is unavailable?**
+###### Are there any tests for feature enablement/disablement?
+Adequate unit tests, component integration test and e2e tests will be added for this feature before
```

Comment:
This was promoted to beta in k/k in https://github.com/kubernetes/kubernetes/pull/117503

Can you please specify what exact tests exist [it says they "will be added [...] before beta" so everything should already be there.

Reply:
that's a great point, i think i'll need to confer with the sig about this. this part was written before i added the new content so i'll need to figure out which tests they were referring to.

Reply:
i have added the enumerated unit tests from the external CCM framework, and linked to the PR which enables them in CI.

should i add a mention of no specific testing of a live upgrade/downgrade?

Reply:
We just overhauled all existing kubernetes CI jobs to now run with the external GCE cloud provider, that is efffectively the test coverage. That plus additional test suites run by each provider downstream. 

We don't necessarily need to link all the test specifically, but maybe add some links to testgrid dashboards if we haven't already. 

We don't have any automated tests that enables and disables this in a test though, for the same reasons we don't have version upgrade/downgrade tests. But as mentioned in other threads, we have signal  from various manual tests and providers who have implemented the the migration downstream (GKE, EKS, AKS, etc)